18 search improvements

patrick-austin opened the pull request:
Covers a variety of improvements to free text search functionality.
Closes #18 #19 #25 #26 #30

Functional changes
--------------------
Added the ability to:
- Search on over 2 billion documents
- Apply sorting on specific entity fields
- "Infinitely" search the data by using the `searchAfter` parameter
- Request the facets of a search
- Expanding the `text` field into more specific fields that reflect the ICAT schema to allow field targetting
- Support for unit conversion on numeric `Parameter`s

Architectural changes
----------------------
In addition to changes to various functions on the main `Lucene` class:
- [ ] Refactored mapping of fields to its own class `DocumentMapping`
- [ ] Created `FacetedDimension` to store the results of a request to facet the documents
- [ ] Refactored the `Search` from `Lucene` into its own dedicated `SearchBucket` class to handle building of complex queries and additional searching parameters (e.g. sorting)
- [ ] `ShardBucket` added alongside `IndexBucket` in `Lucene` to reflect the distinction between a single directory of Lucene Documents (`ShardBucket`) and a way to access all Documents of one entity type contained in one or more directories (`IndexBucket`)

Interdependencies on other components
-----------------------------------------
- icatproject/icat.utils#20
- icatproject/icat.client#33
- icatproject/icat.server#267

VKTB reviewed (changes requested):
There is quite a lot going on here so I must admit that it's hard for me to understand everything but I tried my best.

I can see that you have refactored code into private methods to make some methods shorter but there are still a few methods (i.e. `luceneFacetResult`, `luceneSearchResult` etc in the `Lucene` class) that are long so have a look to see if you can shorten them.

patrick-austin commented on src/main/java/org/icatproject/lucene/SearchBucket.java:
```diff
@@ -0,0 +1,874 @@
+            List<SortField> fields = new ArrayList<>();
+            for (String key : object.keySet()) {
+                String order = object.getString(key);
+                Boolean reverse;
```
I hadn't given thought to primitives vs objects, so there's probably a lot of these changes that could (should) be made.

patrick-austin commented on src/main/java/org/icatproject/lucene/Lucene.java:
```diff
@@ -660,220 +967,754 @@ public void run() {
+	 * @param document Lucene Document being built.
+	 * @param key      Name of the field to be added.
+	 */
+	private void addField(JsonObject json, Document document, String key) {
```
I remember this and `addSortField` being quite painful due to how the Lucene `IndexableField`s behave compared to simple JSON. But there's probably something that can be refactored, I'll have a look another look at it.

patrick-austin commented on src/main/java/org/icatproject/lucene/Lucene.java:
```diff
@@ -660,220 +967,754 @@ public void run() {
+		Document newDocument = new Document();
+		for (IndexableField field : oldDocument.getFields()) {
+			String fieldName = field.name();
+			if (json.keySet().contains(fieldName)) {
```
Can't see any change between the two lines. Is this a case where I can use `containsKey` instead?

VKTB replied:
Ah sorry, I must've pasted the same line.. Yes, you can use `containsKey` here
```suggestion
			if (json.containsKey(fieldName)) {
```

patrick-austin commented on src/main/java/org/icatproject/lucene/Lucene.java:
```diff
@@ -660,220 +967,754 @@ public void run() {
 
+	/**
+	 * Updates an existing Lucene document, provided that the target index is not
+	 * locked
+	 * for another operation.
```
Yeah the automatic formatting VSCode does is really annoying for shuffling one word of a Javadoc onto a new line, thanks for going through this with enough detail to catch that.

patrick-austin commented on src/main/java/org/icatproject/lucene/IcatAnalyzer.java:
```diff
@@ -4,22 +4,33 @@
+// import org.apache.lucene.analysis.standard.StandardAnalyzer ;
 import org.apache.lucene.analysis.standard.StandardTokenizer;
 
+// public class IcatAnalyzer extends Analyzer {
+
+// 	@Override
+// 	protected TokenStreamComponents createComponents(String fieldName) {
+// 		StandardAnalyzer analyzer = new StandardAnalyzer(EnglishAnalyzer.ENGLISH_STOP_WORDS_SET);
+// 		Analyzer.TokenStreamComponents stream = analyzer.createComponents(fieldName);
+// 		sink = new EnglishPossessiveFilter(stream.getTokenStream());
+// 		sink = new PorterStemFilter(sink);
+// 		return new TokenStreamComponents(source, sink);
+// 	}
+// }
+
```
Yeah this shouldn't be here - I think it's removed by #21
Having said that, I probably should ask you to review that too - it's a less complicated, and more self contained change. But if you don't have time I'll see if Kevin can review it himself.

VKTB replied:
I can have a look at it at some point.

patrick-austin commented on src/main/java/org/icatproject/lucene/Lucene.java:
```diff
@@ -295,22 +417,22 @@ private void add(HttpServletRequest request, String entityName, When when, JsonP
+			for (JsonObject document : documents) {
+				createNow(entityName, document);
 			}
+		} catch (JsonException e) {
+			logger.error("Could not parse JSON from {}", value.toString());
```
To be honest I'm not sure. The fact that I've put a logger in there rather than just throwing means it probably did happen when I was developing (and I wanted extra logging to debug it). It's plausible it could happen when we `read()` if the input wasn't a valid `JsonStructure`, or later on when getting the values as Objects. In order to avoid the `NullPointer` could just remove the `toString` like elsewhere, and just let the logger format it?

VKTB replied:
I think I added this comment before noticing the redundant usage of `toString` so I forgot to go back to it and update it, sorry. You are right, I believe that removing the `toString` should address this.

patrick-austin commented:
In addition to the smaller stuff (use of `toString()`, use of primitives, formatting etc.) have:
- Refactored code out of `luceneFacetResult`, `luceneSearchResult` to reduce length (each has two new functions they call)
- Changed more places uses objects instead of primitives (where we parse dates and used to handle default values of `null`, now use `Long.MAX_VALUE` and `Long.MIN_VALUE`
- Replaced `addField` and `addSort` field with a new `Field` class, with equivalent methods. The special treatment that depends on the type of the field and whether it is facetable or sortable is done on an `InnerField` class which are created from either Json or the Lucene IndexableField depending on circumstances.

patrick-austin commented on src/main/java/org/icatproject/lucene/Field.java:
```diff
@@ -0,0 +1,193 @@
+     * @return The original document with this field added to it
+     * @throws NumberFormatException
+     */
+    public Document addToDocument(Document document) throws NumberFormatException {
```
Yeah good point, I suppose I started out by thinking it should return the `Document` like `pruneDocument()` does, but we don't need to here. I guess it depends whether you think `field.addToDocument(document)` or `document = field.addToDocument(document)` looks better in `Lucene.java` (and whether it's clear that the function is going to implicitly change the Document you provide it).

VKTB replied:
I think the former looks better and it's clear enough to me that the passed document will get changed.
